File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/utils/localization/TwoDeadWheelLocalizer.java
package org.firstinspires.ftc.teamcode.utils.localization;

import androidx.annotation.NonNull;

import com.acmerobotics.dashboard.config.Config;
import com.acmerobotics.roadrunner.DualNum;
import com.acmerobotics.roadrunner.Rotation2d;
import com.acmerobotics.roadrunner.Time;
import com.acmerobotics.roadrunner.Twist2dDual;
import com.acmerobotics.roadrunner.Vector2dDual;
import com.acmerobotics.roadrunner.ftc.Encoder;
import com.acmerobotics.roadrunner.ftc.OverflowEncoder;
import com.acmerobotics.roadrunner.ftc.PositionVelocityPair;
import com.acmerobotics.roadrunner.ftc.RawEncoder;
import com.qualcomm.robotcore.hardware.DcMotorEx;
import com.qualcomm.robotcore.hardware.DcMotorSimple;
import com.qualcomm.robotcore.hardware.HardwareMap;
import com.qualcomm.robotcore.hardware.IMU;

import org.firstinspires.ftc.robotcore.external.navigation.AngleUnit;

@Config
public final class TwoDeadWheelLocalizer implements IncrementalLocalizer {
    public static class Params {
        public double parYTicks = -7776.154042686053; // y position of the parallel encoder (in tick units)
        public double perpXTicks = 7886.238360505399; // x position of the perpendicular encoder (in tick units)
    }

    public static Params PARAMS = new Params();

    public final Encoder par, perp;
    public final IMU imu;

    private int lastParPos, lastPerpPos;
    private Rotation2d lastHeading;

    private final double inPerTick;

    private double lastRawHeadingVel, headingVelOffset;

    public TwoDeadWheelLocalizer(HardwareMap hardwareMap, IMU imu, double inPerTick) {
        par = new OverflowEncoder(new RawEncoder(hardwareMap.get(DcMotorEx.class, "leftBack")));
        perp = new OverflowEncoder(new RawEncoder(hardwareMap.get(DcMotorEx.class, "rightBack")));

        par.setDirection(DcMotorSimple.Direction.REVERSE);
        perp.setDirection(DcMotorSimple.Direction.REVERSE);

        this.imu = imu;

        lastParPos = par.getPositionAndVelocity().position;
        lastPerpPos = perp.getPositionAndVelocity().position;
        lastHeading = Rotation2d.exp(imu.getRobotYawPitchRollAngles().getYaw(AngleUnit.RADIANS));

        this.inPerTick = inPerTick;
    }

    // see https://github.com/FIRST-Tech-Challenge/FtcRobotController/issues/617
    private double getHeadingVelocity() {
        double rawHeadingVel = imu.getRobotAngularVelocity(AngleUnit.RADIANS).zRotationRate;
        if (Math.abs(rawHeadingVel - lastRawHeadingVel) > Math.PI) {
            headingVelOffset -= Math.signum(rawHeadingVel) * 2 * Math.PI;
        }
        lastRawHeadingVel = rawHeadingVel;
        return headingVelOffset + rawHeadingVel;
    }

    @NonNull
    public Twist2dDual<Time> update() {
        PositionVelocityPair parPosVel = par.getPositionAndVelocity();
        PositionVelocityPair perpPosVel = perp.getPositionAndVelocity();
        Rotation2d heading = Rotation2d.exp(imu.getRobotYawPitchRollAngles().getYaw(AngleUnit.RADIANS));

        int parPosDelta = parPosVel.position - lastParPos;
        int perpPosDelta = perpPosVel.position - lastPerpPos;
        double headingDelta = heading.minus(lastHeading);

        double headingVel = getHeadingVelocity();

        Twist2dDual<Time> twist = new Twist2dDual<>(
            new Vector2dDual<>(
                    new DualNum<Time>(new double[] {
                        parPosDelta - PARAMS.parYTicks * headingDelta,
                        parPosVel.velocity - PARAMS.parYTicks * headingVel,
                    }).times(inPerTick),
            new DualNum<Time>(new double[] {
                perpPosDelta - PARAMS.perpXTicks * headingDelta,
                perpPosVel.velocity - PARAMS.perpXTicks * headingVel,
            }).times(inPerTick)
        ),
        new DualNum<>(new double[] {
                headingDelta,
                headingVel,
        })
        );

        lastParPos = parPosVel.position;
        lastPerpPos = perpPosVel.position;
        lastHeading = heading;

        return twist;
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/utils/localization/TwoDeadWheelLocalizer.kt
package org.firstinspires.ftc.teamcode.utils.localization

import com.acmerobotics.dashboard.config.Config
import com.acmerobotics.roadrunner.DualNum
import com.acmerobotics.roadrunner.Rotation2d
import com.acmerobotics.roadrunner.Time
import com.acmerobotics.roadrunner.Twist2dDual
import com.acmerobotics.roadrunner.Vector2dDual
import com.acmerobotics.roadrunner.ftc.Encoder
import com.acmerobotics.roadrunner.ftc.FlightRecorder
import com.acmerobotics.roadrunner.ftc.OverflowEncoder
import com.acmerobotics.roadrunner.ftc.RawEncoder
import com.qualcomm.robotcore.hardware.DcMotorEx
import com.qualcomm.robotcore.hardware.DcMotorSimple
import com.qualcomm.robotcore.hardware.HardwareMap
import com.qualcomm.robotcore.hardware.IMU
import org.firstinspires.ftc.robotcore.external.navigation.AngleUnit
import kotlin.math.PI
import kotlin.math.abs
import kotlin.math.sign

@Config
class TwoDeadWheelLocalizer(
    hardwareMap: HardwareMap, private val imu: IMU, private val inPerTick: Double
) : IncrementalLocalizer {
    val par: Encoder
    val perp: Encoder

    private var lastParPos: Int
    private var lastPerpPos: Int
    private var lastHeading: Rotation2d

    private var lastRawHeadingVel = 0.0
    private var headingVelOffset = 0.0

    init {
        par = OverflowEncoder(RawEncoder(hardwareMap.get(DcMotorEx::class.java, "rightFront")))
        perp = OverflowEncoder(RawEncoder(hardwareMap.get(DcMotorEx::class.java, "rightBack")))

        par.direction = DcMotorSimple.Direction.REVERSE
        perp.direction = DcMotorSimple.Direction.REVERSE

        lastParPos = par.getPositionAndVelocity().position
        lastPerpPos = perp.getPositionAndVelocity().position
        lastHeading = Rotation2d.exp(imu.robotYawPitchRollAngles.getYaw(AngleUnit.RADIANS))

        FlightRecorder.write("TWO_DEAD_WHEEL_PARAMS", PARAMS)
    }

    private fun getHeadingVelocity(): Double {
        val rawHeadingVel = imu.getRobotAngularVelocity(AngleUnit.RADIANS).zRotationRate.toDouble()
        if (abs(rawHeadingVel - lastRawHeadingVel) > PI) {
            headingVelOffset -= sign(rawHeadingVel) * 2 * PI
        }
        lastRawHeadingVel = rawHeadingVel
        return headingVelOffset + rawHeadingVel
    }

    override fun update(): Twist2dDual<Time> {
        val parPosVel = par.getPositionAndVelocity()
        val perpPosVel = perp.getPositionAndVelocity()
        val heading: Rotation2d =
            Rotation2d.exp(imu.robotYawPitchRollAngles.getYaw(AngleUnit.RADIANS))

        val parPosDelta = parPosVel.position - lastParPos
        val perpPosDelta = perpPosVel.position - lastPerpPos
        val headingDelta = heading.minus(lastHeading)

        val headingVel = getHeadingVelocity()

        val twist = Twist2dDual(
            Vector2dDual(
                DualNum<Time>(
                    listOf(
                        parPosDelta - PARAMS.parYTicks * headingDelta,
                        parPosVel.velocity - PARAMS.parYTicks * headingVel
                    )
                ).times(inPerTick),
                DualNum<Time>(
                    listOf(
                        perpPosDelta - PARAMS.perpXTicks * headingDelta,
                        perpPosVel.velocity - PARAMS.perpXTicks * headingVel
                    )
                ).times(inPerTick)
            ),
            DualNum(
                listOf(
                    headingDelta,
                    headingVel
                )
            )
        )

        lastParPos = parPosVel.position
        lastPerpPos = perpPosVel.position
        lastHeading = heading

        return twist
    }

    companion object {
        class Params {
            var parYTicks =
                7915.166024327065 // y position of the parallel encoder (in tick units)
            var perpXTicks =
                7592.816187212765 // x position of the perpendicular encoder (in tick units)
        }

        @JvmStatic
        var PARAMS = Params()
    }
}